// discord-giveaway-easy/src/CompleteGiveaway/utils/Constants.js
const Discord = require("discord.js")




/**
 * @typedef GiveawaysStartOptions
 * 
 * @property {String} [time]
 * @property {Number} [winnerNumber]
 * @property {Discord.ButtonStyle} [buttonType]
 * @property {Discord.Message} [interaction]
 * @property {String} [prize]
 */

exports.GiveawaysStartOptions = {}






/**
 * @typedef GiveawaysManagerOptions
 * 
 * @property {String} [embedColor]
 * @property {String} [buttonEmoji]
 *
 */

 exports.GiveawaysManagerOptions = {}

 





/**
 * @typedef EmbedGeneratorOptions
 * 
 * @property {Number} [winnerNumber]
 * @property {Discord.Message} [interaction]
 * @property {String} [Title]
 * @property {String} [Color]
 * @property {String} [Emoji]
 * @property {String} [winnerRandom]
 * @property {String} [entries]
 * @property {Discord.Message} [messageEditable]
 * 
 */
 exports.EmbedGeneratorOptions = {};


/**
 * @typedef GiveawayOn
 * 
 * @property {Discord.Client} [client]
 * @property {Number} [time]
 * @property {Discord.Message} [interaction]
 */

exports.GiveawayOn = {};


 
/**
 * @typedef MessageGiveaways
 * 
 * @property {String} [giveaway="🎉🎉 **GIVEAWAY** 🎉🎉"]
 * @property {String} [giveawayEnded="🎉🎉 **GIVEAWAY FINI** 🎉🎉"]
 * @property {String} [giveawayStoped="🎉🎉 **GIVEAWAY ARRETE** 🎉🎉"]
 * 
 */
 exports.MessageGiveaways = {
    giveaway: '🎉🎉 **GIVEAWAY** 🎉🎉',
    giveawayEnded: '🎉🎉 **GIVEAWAY FINI** 🎉🎉',
    giveawayStoped: '🎉🎉 **GIVEAWAY ARRETE** 🎉🎉'
 }


 /**
 * @typedef CreateInteraction
 * 
 * @property {Discord.ButtonStyle} [buttonStyle]
 * @property {String} [buttonEmoji]
 * @property {Number} [winnerNumber]
 * @property {Discord.Client} [client]
 * 
 */
  exports.CreateInteraction = {}
