// discord-giveaway-easy/src/CompleteGiveaway/utils/EmbedGeneratorBuilder.js
const Discord = require("discord.js")
const {EmbedGeneratorOptions, EmbedEndGeneratorOptions} = require("./Constants")

module.exports = class EmbedGeneratorBuilder {
  /**
   * @param {Number} time
   * @param {EmbedGeneratorOptions} options
   * 
   * @exemple
   * ```
   * let time = "10m"
   * 
   * new EmbedGeneratorBuilder(time, {
   *  interaction: message, 
   *  winnerNumber: options.winnerNumber,
   *  Title: "TitleExemple",
   *  Color: "#ffffff",
   * })
   * ```
   */
  constructor(time, options) {
    
    this.time = time;
    this.emoji = options.Emoji;
    this.winnerNumber = options.winnerNumber;
    this.interaction = options.interaction;
    this.Title = options.Title;
    this.Color = options.Color;
    this.winnerRandom = options.winnerRandom;

    this.entries = options.entries;
  }
 
   Embed() {
    let Embed = new Discord.EmbedBuilder()
    .setTitle(this.Title)
    .setColor(this.Color)
    .setDescription(`Réagit avec ${this.emoji} pour participer!\nTemps restant: <t:${this.time}:R>\nLancé par: ${this.interaction.user}`)
    .setTimestamp()
    .setFooter({text: `gagnant(s): ${this.winnerNumber}`})
    
    return Embed;
  }

  /**
   * 
   * @param {Number} rerollTime 
   * @returns 
   */
   endEmbed(rerollTime) {
    let Embed = new Discord.EmbedBuilder()
       .setTitle(this.Title)
       .setColor("#2f3136")
       .setDescription(`Fin: <t:${rerollTime === undefined ? this.time : this.time + rerollTime}:R> (<t:${this.time}:f>)\nGagnant(s): \n${this.winnerRandom.join("\n")}\nlancé par: ${this.interaction.user}`)
       .setFooter({text: `Gagnant(s): ${this.winnerNumber}`})
       .setTimestamp()

       return Embed;  
  }

  /**
   * 
   * @param {Discord.Message} message 
   * @returns 
   */
  entrieEmbed(message) {
    let EntrieEmbed = new Discord.EmbedBuilder()
        .setDescription(`**${this.entries}** entrées. [↗️](${message.url})`)
        .setColor("#2f3136")

        return EntrieEmbed;
  }

  
}
